File: SNPsDensity-LOH_index/SNPsDensityAnalisys/bcftools/script/orc_analyzer.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <filesystem>
#include <climits>

using SNPMap = std::unordered_map<std::string, std::vector<int>>;

// Structure representing a genomic region (e.g., from a BED file)
struct Region {
    std::string chrom;   // Chromosome name
    int start;           // Start coordinate (0-based)
    int end;             // End coordinate
    std::string id;      // Unique identifier
    std::string strand;  // Strand information (+, -, or .)
};

// Reads chromosome names and their total lengths from a file
std::unordered_map<std::string, int> read_chrom_sizes(const std::string& path) {
    std::unordered_map<std::string, int> chrom_sizes;
    std::ifstream infile(path);
    std::string chrom;
    int size;
    while (infile >> chrom >> size) {
        chrom_sizes[chrom] = size;
    }
    return chrom_sizes;
}

// Reads SNP positions from multiple files and sorts them for efficient searching
SNPMap read_snps(const std::vector<std::string>& snpFiles) {
    SNPMap snps;
    for (const auto& path : snpFiles) {
        std::ifstream infile(path);
        std::string chrom;
        int pos;
        while (infile >> chrom >> pos) {
            snps[chrom].push_back(pos);
        }
    }
    // Sorting allows for binary search (O(log n)) instead of linear search
    for (auto& [chr, positions] : snps) {
        std::sort(positions.begin(), positions.end());
    }
    return snps;
}

// Reads genomic regions from a BED-formatted file
std::vector<Region> read_bed(const std::string& path) {
    std::vector<Region> regions;
    std::ifstream infile(path);
    std::string line;
    while (std::getline(infile, line)) {
        std::istringstream iss(line);
        std::string chrom, strand, score;
        int start, end;
        // Expecting standard BED columns
        if (!(iss >> chrom >> start >> end >> strand >> score >> strand)) continue;
        regions.push_back({chrom, start, end, "", strand});
    }
    return regions;
}

// Checks which strand a specific window belongs to based on a reference GFF/BED
std::string get_strand(const std::vector<Region>& genome_bed, const std::string& chrom, int start, int end) {
    for (const auto& region : genome_bed) {
        if (region.chrom == chrom && start >= region.start && end <= region.end) {
            return region.strand;
        }
    }
    return "GAP"; // Returns GAP if the region is not covered in the reference
}

// Counts SNPs within a range [start, end) using binary search for performance
int count_snps(const SNPMap& snps, const std::string& chrom, int start, int end) {
    if (snps.find(chrom) == snps.end()) return 0;
    const auto& positions = snps.at(chrom);
    auto it_start = std::lower_bound(positions.begin(), positions.end(), start);
    auto it_end = std::lower_bound(positions.begin(), positions.end(), end);
    return std::distance(it_start, it_end);
}

// Formats and writes the calculated data for a single region/window to the output file
void write_region(std::ofstream& out, const std::string& id, const std::string& sample, const std::string& region_type,
                  const std::string& chrom, int start, int end, const std::string& group,
                  const std::string& strand, int snp_count) {
    int length = end - start;
    // SNP density is calculated as $Density = \frac{count}{length}$
    double snp_density = (length > 0) ? static_cast<double>(snp_count) / length : 0.0;

    out << id << "\t" << sample << "\t" << region_type << "\t" << chrom << "\t"
    << start << "\t" << end << "\t" << length << "\t" << snp_count << "\t"
    << snp_density << "\t" << group << "\t" << strand << "\n";
                  }

                  // Generates 2000bp windows up to a 30kb limit around a target site
                  void create_windows(std::ofstream& out, const SNPMap& snps, const std::vector<Region>& genome_bed,
                                      const std::unordered_map<std::string, int>& chrom_sizes,
                                      const Region& site, const std::string& sample_name, const std::string& group) {

                      int start = site.start;
                      int end = site.end;
                      int chr_limit = chrom_sizes.at(site.chrom);
                      std::string id = "ORC_" + site.chrom + "_" + std::to_string(start) + "_" + std::to_string(end) + "_" + group;

                      const int WINDOW_SIZE = 2000;
                      const int MAX_DISTANCE = 30000; // Limit counting to 30kb from the site

                      // 1. Process the central site itself
                      int site_snps = count_snps(snps, site.chrom, start, end);
                      std::string site_strand = get_strand(genome_bed, site.chrom, start, end);
                      write_region(out, id, sample_name, "site", site.chrom, start, end, group, site_strand, site_snps);

                      // 2. Process Upstream regions (moving backwards from 'start')
                      for (int offset = WINDOW_SIZE; offset <= MAX_DISTANCE; offset += WINDOW_SIZE) {
                          int w_start = start - offset;
                          int w_end = start - (offset - WINDOW_SIZE);

                          if (w_start < 0) {
                              // Handle cases near the beginning of the chromosome
                              if (w_end > 0) {
                                  int count = count_snps(snps, site.chrom, 0, w_end);
                                  std::string strand = get_strand(genome_bed, site.chrom, 0, w_end);
                                  write_region(out, id, sample_name, "upstream_" + std::to_string(offset) + "bp", site.chrom, 0, w_end, group, strand, count);
                              }
                              break; // Stop if we reach the start of the chromosome
                          } else {
                              int count = count_snps(snps, site.chrom, w_start, w_end);
                              std::string strand = get_strand(genome_bed, site.chrom, w_start, w_end);
                              write_region(out, id, sample_name, "upstream_" + std::to_string(offset) + "bp", site.chrom, w_start, w_end, group, strand, count);
                          }
                      }

                      // 3. Process Downstream regions (moving forwards from 'end')
                      for (int offset = WINDOW_SIZE; offset <= MAX_DISTANCE; offset += WINDOW_SIZE) {
                          int w_start = end + (offset - WINDOW_SIZE);
                          int w_end = end + offset;

                          if (w_end > chr_limit) {
                              // Handle cases near the end of the chromosome
                              if (w_start < chr_limit) {
                                  int count = count_snps(snps, site.chrom, w_start, chr_limit);
                                  std::string strand = get_strand(genome_bed, site.chrom, w_start, chr_limit);
                                  write_region(out, id, sample_name, "downstream_" + std::to_string(offset) + "bp", site.chrom, w_start, chr_limit, group, strand, count);
                              }
                              break; // Stop if we reach the end of the chromosome
                          } else {
                              int count = count_snps(snps, site.chrom, w_start, w_end);
                              std::string strand = get_strand(genome_bed, site.chrom, w_start, w_end);
                              write_region(out, id, sample_name, "downstream_" + std::to_string(offset) + "bp", site.chrom, w_start, w_end, group, strand, count);
                          }
                      }
                                      }

                                      int main(int argc, char* argv[]) {
                                          if (argc < 6) {
                                              std::cerr << "Usage: " << argv[0] << " <chromSizes> <GFF_BED> <BED_DIR> <OUTPUT_FILE> <SNP_TSV_FILES...>" << std::endl;
                                              return 1;
                                          }

                                          // Initialize file paths from command line arguments
                                          std::string chrom_sizes_file = argv[1];
                                          std::string gff_file = argv[2];
                                          std::string bed_dir = argv[3];
                                          std::string output_file = argv[4];
                                          std::vector<std::string> snp_files(argv + 5, argv + argc);

                                          // Load data into memory
                                          auto chrom_sizes = read_chrom_sizes(chrom_sizes_file);
                                          auto snps = read_snps(snp_files);
                                          auto genome_bed = read_bed(gff_file);

                                          // Prepare output file and write header
                                          std::ofstream outfile(output_file);
                                          outfile << "ID\tsample\tregion\tchrom\tstart\tend\tlength\tSNPs_count\tSNPs_density\tGroup\tStrand\n";

                                          // Iterate through all BED files in the specified directory
                                          for (const auto& bed_file : std::filesystem::directory_iterator(bed_dir)) {
                                              auto regions = read_bed(bed_file.path().string());
                                              std::string group = bed_file.path().stem().string();

                                              for (const auto& site : regions) {
                                                  // Processing for 4 hypothetical samples as per original logic
                                                  for (int i = 1; i <= 4; ++i) {
                                                      create_windows(outfile, snps, genome_bed, chrom_sizes, site, "sample" + std::to_string(i), group);
                                                  }
                                              }
                                          }

                                          return 0;
                                      }

File: SNPsDensity-LOH_index/SNPsDensityAnalisys/bcftools/script/orc_analyzer.py
#!/usr/bin/env python3
import sys
import bisect
from pathlib import Path
from dataclasses import dataclass
from typing import Dict, List

# --- Data Structures ---

@dataclass
class Region:
    chrom: str
    start: int
    end: int
    strand: str

# --- File Reading Functions ---

def read_chrom_sizes(path: str) -> Dict[str, int]:
    """Reads chromosome sizes file (format: chrom \t size)."""
    chrom_sizes = {}
    try:
        with open(path, 'r') as f:
            for line in f:
                parts = line.strip().split("\t")
                if len(parts) >= 2:
                    chrom_sizes[parts[0]] = int(parts[1])
    except Exception as e:
        print(f"Error reading chromosome sizes: {e}")
        sys.exit(1)
    return chrom_sizes

def read_bed_to_dict(path: str) -> Dict[str, List[Region]]:
    """Reads a BED file and organizes regions by chromosome for fast lookup."""
    data = {}
    try:
        with open(path, 'r') as f:
            for line in f:
                if line.startswith('#') or not line.strip():
                    continue
                parts = line.strip().split()
                if len(parts) >= 6:
                    chrom = parts[0]
                    if chrom not in data:
                        data[chrom] = []
                    data[chrom].append(Region(
                        chrom=chrom,
                        start=int(parts[1]),
                        end=int(parts[2]),
                        strand=parts[5]
                    ))
    except Exception as e:
        print(f"Warning reading {path}: {e}")
    return data

def read_snps_by_sample(snp_files: List[str]) -> Dict[str, Dict[str, List[int]]]:
    """
    Reads multiple SNP files and stores them per sample and per chromosome.
    Returns: { 'sample1': { 'chr1': [pos1, pos2], ... }, 'sample2': ... }
    """
    all_samples_snps = {}
    for i, path in enumerate(snp_files):
        sample_name = f"sample{i+1}"
        all_samples_snps[sample_name] = {}
        try:
            with open(path, 'r') as f:
                for line in f:
                    parts = line.strip().split()
                    if len(parts) >= 2:
                        chrom, pos = parts[0], int(parts[1])
                        if chrom not in all_samples_snps[sample_name]:
                            all_samples_snps[sample_name][chrom] = []
                        all_samples_snps[sample_name][chrom].append(pos)

            # Sorting is essential for binary search (O(log N))
            for chrom in all_samples_snps[sample_name]:
                all_samples_snps[sample_name][chrom].sort()
            print(f"[*] Loaded SNPs for {sample_name} from {path}")
        except Exception as e:
            print(f"Error loading SNP file {path}: {e}")
    return all_samples_snps

# --- Processing Logic ---

def get_strand_from_poly(chrom: str, start: int, end: int, poly_ref: Dict[str, List[Region]]) -> str:
    """
    Assigns strand if the window is fully contained within a polycistron.
    Otherwise returns 'GAP'.
    """
    if chrom in poly_ref:
        for poly in poly_ref[chrom]:
            # Rule: Window must be strictly inside the polycistron limits
            if start >= poly.start and end <= poly.end:
                return poly.strand
    return "GAP"

def count_snps(sample_snps: Dict[str, List[int]], chrom: str, start: int, end: int) -> int:
    """Counts SNPs in range [start, end) using efficient binary search."""
    if chrom not in sample_snps:
        return 0
    positions = sample_snps[chrom]
    idx_start = bisect.bisect_left(positions, start)
    idx_end = bisect.bisect_left(positions, end)
    return idx_end - idx_start

def write_line(out_f, region_id, sample, r_type, chrom, start, end, group, strand, snp_count):
    """Calculates density and writes a tab-separated line to the output file."""
    length = end - start
    density = (snp_count / length) if length > 0 else 0.0
    line = (f"{region_id}\t{sample}\t{r_type}\t{chrom}\t"
            f"{start}\t{end}\t{length}\t{snp_count}\t"
            f"{density:.6f}\t{group}\t{strand}\n")
    out_f.write(line)

def process_windows(out_f, sample_name, sample_snps, site, poly_ref, chrom_sizes, group):
    """Generates the central site and 30kb flanking windows (upstream/downstream)."""
    start, end = site.start, site.end
    chrom = site.chrom
    chr_limit = chrom_sizes.get(chrom, 0)
    region_id = f"ORC_{chrom}_{start}_{end}_{group}"

    WINDOW_SIZE = 2000
    MAX_DISTANCE = 30000

    # 1. Central Site
    s_strand = get_strand_from_poly(chrom, start, end, poly_ref)
    s_count = count_snps(sample_snps, chrom, start, end)
    write_line(out_f, region_id, sample_name, "site", chrom, start, end, group, s_strand, s_count)

    # 2. Upstream Flanks (15 windows of 2kb)
    for offset in range(WINDOW_SIZE, MAX_DISTANCE + 1, WINDOW_SIZE):
        w_start = max(0, start - offset)
        w_end = start - (offset - WINDOW_SIZE)
        if w_end <= 0: break # Out of chromosome bounds

        u_count = count_snps(sample_snps, chrom, w_start, w_end)
        u_strand = get_strand_from_poly(chrom, w_start, w_end, poly_ref)
        write_line(out_f, region_id, sample_name, f"upstream_{offset}bp", chrom, w_start, w_end, group, u_strand, u_count)

    # 3. Downstream Flanks (15 windows of 2kb)
    for offset in range(WINDOW_SIZE, MAX_DISTANCE + 1, WINDOW_SIZE):
        w_start = end + (offset - WINDOW_SIZE)
        w_end = min(chr_limit, end + offset)
        if w_start >= chr_limit: break # Out of chromosome bounds

        d_count = count_snps(sample_snps, chrom, w_start, w_end)
        d_strand = get_strand_from_poly(chrom, w_start, w_end, poly_ref)
        write_line(out_f, region_id, sample_name, f"downstream_{offset}bp", chrom, w_start, w_end, group, d_strand, d_count)

# --- Main Execution ---

def main():
    if len(sys.argv) < 6:
        print("Usage: python3 orc_analyzer.py <chromSizes> <Poly_BED> <Input_BED_Dir_or_File> <Output_TSV> <SNP_File1> <SNP_File2> ...")
        sys.exit(1)

    # Input arguments
    chrom_sizes_path = sys.argv[1]
    poly_bed_path = sys.argv[2]
    input_path = Path(sys.argv[3])
    output_path = sys.argv[4]
    snp_files = sys.argv[5:] # Files for sample1, sample2, etc.

    print("[*] Initializing analysis...")

    # Load reference data
    chrom_sizes = read_chrom_sizes(chrom_sizes_path)
    poly_ref = read_bed_to_dict(poly_bed_path)
    all_snps_data = read_snps_by_sample(snp_files)

    # Determine which BED files to process (Origins, Initiation Zones, etc.)
    if input_path.is_dir():
        bed_files = sorted(list(input_path.glob("*.bed")))
    else:
        bed_files = [input_path]

    # Process and write results
    with open(output_path, 'w') as out_file:
        # TSV Header
        out_file.write("ID\tsample\tregion\tchrom\tstart\tend\tlength\tSNPs_count\tSNPs_density\tGroup\tStrand\n")

        for b_file in bed_files:
            group_name = b_file.stem
            print(f"[*] Processing group: {group_name}")

            # Load ORC sites for the current group
            sites_to_analyze = read_bed_to_dict(str(b_file))

            for chrom, sites in sites_to_analyze.items():
                for site in sites:
                    # Iterate through each sample to ensure independent SNP counting
                    for sample_name, sample_snps in all_snps_data.items():
                        process_windows(out_file, sample_name, sample_snps, site, poly_ref, chrom_sizes, group_name)

    print(f"[*] Done! Results saved to: {output_path}")

if __name__ == "__main__":
    main()
